### PR TITLE
[mtl-007] AEC fixes from main branch submission

### PR DESCRIPTION
**Pull request by andyross**
Squashed changes from https://github.com/thesofproject/sof/pull/8571 , rebased on top of the mtl-007-drop-stable branch so we can fix the divergences in this code.

Review there.  Not here.

### CODE REVIEW

**Review by marcinszkudlinski — CHANGES_REQUESTED** [review]
007 branch is a frozen release "stable" branch for MTL, please don't merge this commit.

File: app/boards/intel_adsp_ace15_mtpm.conf [hunk 1]
```diff
@@ -94,4 +94,7 @@ CONFIG_PROBE=y
 CONFIG_PROBE_DMA_MAX=2
 CONFIG_KCPS_DYNAMIC_CLOCK_CONTROL=y
 CONFIG_COMP_GOOGLE_RTC_AUDIO_PROCESSING=y
+CONFIG_COMP_GOOGLE_RTC_AUDIO_PROCESSING_NUM_CHANNELS=2
+CONFIG_GOOGLE_RTC_AUDIO_PROCESSING_MIC_BITS=32
+
```
Comment (btian1) [inline app/boards/intel_adsp_ace15_mtpm.conf:99]: should no space line?

Comment (andyross) [reply app/boards/intel_adsp_ace15_mtpm.conf:99]: Please review in #8571 . This PR is just a place to deliver the same code against mtl-007-drop-stable.

Comment (andyross) [reply app/boards/intel_adsp_ace15_mtpm.conf:99]: Oh, actually this patch got dropped from that branch somehow!  Will fix.  But when I do please review there :)

File: src/audio/module_adapter/module_adapter.c [hunk 2]
```diff
@@ -1088,9 +1088,6 @@ static int module_adapter_copy_dp_queues(struct comp_dev *dev)
 		struct sof_source *data_src = dp_queue_get_source(dp_queue);
 		size_t dp_data_available = source_get_data_available(data_src);
 
-		if (!dp_data_available)
```
Comment (marcinszkudlinski) [inline src/audio/module_adapter/module_adapter.c:1091]: this error is a real error - if DP processing is working it will never appear, even during pipelie startup

File: tools/topology/topology2/include/components/google-rtc-aec.conf [hunk 3]
```diff
@@ -47,6 +47,15 @@ Class.Widget."google-rtc-aec" {
 		token_ref	"comp.word"
 	}
 
+	Object.Control.bytes."1" {
+		access [
+			tlv_write
+			tlv_callback
```
Comment (cujomalainey) [inline tools/topology/topology2/include/components/google-rtc-aec.conf:53]: is callback a kernel or userspace permission?

File: src/audio/google/google_rtc_audio_processing.c [hunk 4]
```diff
@@ -53,21 +52,48 @@ DECLARE_SOF_RT_UUID("google-rtc-audio-processing", google_rtc_audio_processing_u
...
+
+static __aligned(PLATFORM_DCACHE_ALIGN)
+uint8_t aec_mem_blob[CONFIG_COMP_GOOGLE_RTC_AUDIO_PROCESSING_MEMORY_BUFFER_SIZE_BYTES];
+
+#define NUM_FRAMES (CONFIG_COMP_GOOGLE_RTC_AUDIO_PROCESSING_SAMPLE_RATE_HZ \
+		    / GOOGLE_RTC_AUDIO_PROCESSING_FREQENCY_TO_PERIOD_FRAMES)
+#define CHAN_MAX CONFIG_COMP_GOOGLE_RTC_AUDIO_PROCESSING_CHANNEL_MAX
+
+static __aligned(PLATFORM_DCACHE_ALIGN)
+float refoutbuf[CHAN_MAX][NUM_FRAMES];
```
Comment (cujomalainey) [inline src/audio/google/google_rtc_audio_processing.c:71]: not a fan of having these here long term. Should we want to start supporting >1 AEC instance (e.g. on HS) this will not end well.

File: src/audio/google/google_rtc_audio_processing.c [hunk 5]
```diff
@@ -80,28 +106,130 @@ void GoogleRtcFree(void *ptr)
 	return rfree(ptr);
 }
 
-static void google_rtc_audio_processing_params(struct processing_module *mod,
-					       struct sof_source *ref,
-					       struct sof_source *mic,
-					       struct sof_sink *out)
+static ALWAYS_INLINE float s16_to_float(const char *ptr)
```
Comment (cujomalainey) [inline src/audio/google/google_rtc_audio_processing.c:109]: Why are we not using the macros [here](https://github.com/thesofproject/sof/blob/b3b9abffe7ee85bf0884d392d3d243161efadbbe/src/include/sof/audio/format.h#L71)?

Comment (andyross) [reply src/audio/google/google_rtc_audio_processing.c:109]: (Again, please continue review in the main PR, but answering here):

This was discussed earlier: Those generate horrifyingly pessimal code (runtime division!) and can't be fixed without breaking the ABI (because as specified they work for 64 bit values).  These seem like they're intended for generating compile time constants for fixed point use.  There's also a pcm_convert API elsewhere in SOF, but that's designed to work in a block on linear arrays of samples (and don't expose inlinable single-sample conversions) where here we need to unify de/interleaving with the conversion.

Basically: SOF just doesn't have an API for this.  I could move this somewhere else and try to unify with an existing library.

File: tools/topology/topology2/include/components/google-rtc-aec.conf (hunk 3 above)
Comment (andyross) [reply tools/topology/topology2/include/components/google-rtc-aec.conf:53]: This is a cherry pick from main.  And "access" in this context refers (I think!) to the API variant used to access the control's data.  "callback" is what needs to be used for large TLV blocks because they can't be sent in a single IPC command.

**Comment by andyross** [comment]
Updated with current versions of all my in-review PRs, unifying the AEC work in main with the MTL branch.

**Comment by andyross** [comment]
Close this one.   Main has now caught up with 007-drop-stable and this works directly on top of it.  No need to maintain the separate branch anymore.
